pydoctor/epydoc/markup/_types.py

```diff
@@ -0,0 +1,188 @@
+                    else:
+                        self.tokens.append(node)
+                        raise nodes.SkipNode()
+    
+        tokenizer = Tokenizer(spec)
+        spec.walk(tokenizer)
+        self.warnings.extend(tokenizer.warnings)
+        return tokenizer.tokens
+
+    def _convert_obj_tokens_to_stan(self, tokens: List[Tuple[Union[str, nodes.Node], TokenType]], 
+                                    docstring_linker: DocstringLinker) -> List[Tuple[Union[str, Tag, nodes.Node], TokenType]]:
+        """
+        Convert L{TokenType.OBJ} and PEP 484 like L{TokenType.DELIMITER} type to stan, merge them together. Leave the rest untouched. 
+
+        Exemple:
+
+        >>> tokens = [("list", TokenType.OBJ), ("(", TokenType.DELIMITER), ("int", TokenType.OBJ), (")", TokenType.DELIMITER)]
+        >>> ann._convert_obj_tokens_to_stan(tokens, NotFoundLinker())
+        ... [(Tag('code', children=['list', '(', 'int', ')']), TokenType.OBJ)]
+        
+        @param tokens: List of tuples: C{(token, type)}
+        """
+
+        combined_tokens: List[Tuple[Union[str, Tag], TokenType]] = []
+
+        open_parenthesis = 0
+        open_square_braces = 0
+
+        for _token, _type in tokens:
+
+            if _type is TokenType.OBJ:
+                new_token = docstring_linker.link_xref(_token, _token, self._lineno)
+                if open_square_braces + open_parenthesis > 0:
+                    try: last_processed_token = combined_tokens[-1]
+                    except IndexError: 
+                        # weird
+                        combined_tokens.append((_token, _type))
+                    else:
+                        if last_processed_token[1] is TokenType.OBJ and isinstance(last_processed_token[0], Tag):
+                            # Merge with last Tag
+                            last_processed_token[0](*new_token.children)
+                        else:
+                            # weird
+                            combined_tokens.append((new_token, _type))
+                else:
+                    combined_tokens.append((new_token, _type))
+
+            elif _type is TokenType.DELIMITER: 
+                if _token == "[": open_square_braces += 1
+                elif _token == "(": open_parenthesis += 1
+
+                if open_square_braces + open_parenthesis > 0:
+                    try: last_processed_token = combined_tokens[-1]
+                    except IndexError: 
+                        # weird
+                        combined_tokens.append((_token, _type))
+                    else:
+                        if last_processed_token[1] is TokenType.OBJ and isinstance(last_processed_token[0], Tag): 
+                            # Merge with last Tag
+                            last_processed_token[0](_token)
+                        else:
+                            # weird
+                            combined_tokens.append((_token, _type))
```

Comment:
This corner case seem less difficult to test, this could happen for instance if one tries to convert a type like:
```
:type: default[str]
```

Because `default` is a special keyword. As well as `optional`. I'm not 100% sure this will get this block to be tested, though. This should render like *default*[`str`] and NOT `default[str]`, I suppose. 

Also the following types could make this block to execute, because the first token will be considered as a natural language delimiter, literal, reference or unknown not an object:

```
:type: ,[str]
:type: , [str]
:type: of [str]
:type: or [str]
:type: : [str]
:type: and [str]
```

```
:type: 'hello'[str]
:type: "hello"[str]
```

```
:type: `hello`[str]
:type: `hello <https://github.com>`_[str]
```

```
:type: **hello**[str]
```

I think we should leave a warning for those types.